### PR TITLE
InMemory implementation - Nick Karpov

### PR DESCRIPTION
**Pull request by nickkarpov**
One thing is I had failed tests (tracking for memory leaks) when run on GitHub environment - https://github.com/essentialdevelopercom/ios-lead-essentials-feed-store-challenge/pull/92/checks?sha=177ac3623e8a9e883f82e13899c4e3af6b0505be
It is related to capturing self as strong on operationsQueue that owned by self. I wondering why they always pass locally on my machine?

### CODE REVIEW

File: FeedStoreChallenge/InMemoryFeedStore.swift [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+//
+//  Created by Nick on 19/02/2021.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import Foundation
+
+public class InMemoryFeedStore {
+	private var cachedFeed = [LocalFeedImage]()
+	private var lastUpdate = Date()
```
Comment (caiozullo) [inline FeedStoreChallenge/InMemoryFeedStore.swift:13]: The `cachedFeed ` and `lastUpdate` shouldn't have default values.

It'd be best to combine these two values into a new type or a simple tuple. 

This way, you can perform the operation in one step and also represent the `empty` cache with an optional:

```
private var cache: (feed: [LocalFeedImage], timestamp: Date)?
```

File: FeedStoreChallenge/InMemoryFeedStore.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+			self?.cachedFeed = feed
+			self?.lastUpdate = timestamp
+			completion(nil)
+		}
+	}
+
+	public func retrieve(completion: @escaping RetrievalCompletion) {
+		operationsQueue.async { [cachedFeed, lastUpdate] in
+			if cachedFeed.isEmpty {
+				completion(.empty)
```
Comment (caiozullo) [inline FeedStoreChallenge/InMemoryFeedStore.swift:41]: An empty array of images is a valid value in the cache - so this class shouldn't check for `isEmpty`. 

This class should complete with `.empty` only when the cache is indeed empty (no value is stored). See my other comment on how to represent it with an optional instead.

**Review by caiozullo — COMMENTED** [review]
Hi @nickkarpov, you almost passed the challenge.

Please, check my review comments and let me know once you're done so I can review it again. Thank you 🙂

**Review by caiozullo — COMMENTED** [review]
Hi @nickkarpov,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅ 
- You broke down the implementation into small commits ✅ 
- Bonus points for the Memory Leak Tracking ✅ 

Great job!

**Comment by caiozullo** [comment]
Hi @nickkarpov, regarding your question about the memory leak tracking: 

If the queue closure retains `self` for longer than the test lifetime, it'll fail. But it can fail intermittently since the queue may operation may finish before sometimes, and after the test method finishes other times (the CI serve may be slower, so it may happen there more often). 

So it's important to fix it to have reliable tests regardless of the environment.

**Comment by nickkarpov** [comment]
> Hi @nickkarpov, regarding your question about the memory leak tracking:
> 
> If the queue closure retains `self` for longer than the test lifetime, it'll fail. But it can fail intermittently since the queue may operation may finish before sometimes, and after the test method finishes other times (the CI serve may be slower, so it may happen there more often).
> 
> So it's important to fix it to have reliable tests regardless of the environment.

Thanks Caio,
I also though it could depend on the slower environment on the ci. I just wanted to make sure I can trust my memory leaks tests )

**Pull request closed without merging**